File: minigit-4/py/commands/hash_object.py
"""hash-object command - Compute object hash."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root
from objects import hash_object


def run(args):
    """Compute object hash."""
    # Parse arguments
    write = False
    obj_type = 'blob'
    files = []

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-w':
            write = True
        elif arg == '-t' and i + 1 < len(args):
            obj_type = args[i + 1]
            i += 1
        elif not arg.startswith('-'):
            files.append(arg)
        i += 1

    if not files:
        print("fatal: no files specified", file=sys.stderr)
        return 1

    repo_root = None
    if write:
        repo_root = find_repo_root()
        if repo_root is None:
            print("fatal: not a minigit repository", file=sys.stderr)
            return 1

    for file_path in files:
        try:
            content = Path(file_path).read_bytes()
            sha = hash_object(content, obj_type, write=write, repo_root=repo_root)
            print(sha)
        except FileNotFoundError:
            print(f"fatal: could not open '{file_path}'", file=sys.stderr)
            return 1

    return 0

File: minigit-4/ts/src/commands/hash-object.ts
// hash-object command - Compute object hash

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot } from '../utils';
import { createBlobContent, hashObject, writeObject } from '../objects';

export function hashObjectCmd(args: string[]): number {
  let write = false;
  let objectType = 'blob';
  let filePath: string | null = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '-w') {
      write = true;
    } else if (args[i] === '-t' && i + 1 < args.length) {
      objectType = args[++i];
    } else if (!args[i].startsWith('-')) {
      filePath = args[i];
    }
  }

  if (!filePath) {
    console.error('fatal: file path required');
    return 1;
  }

  const absolutePath = path.resolve(filePath);

  if (!fs.existsSync(absolutePath)) {
    console.error(`fatal: could not open '${filePath}' for reading: No such file or directory`);
    return 1;
  }

  const content = fs.readFileSync(absolutePath);

  // Currently only blob type is fully supported
  let objectContent: Buffer;
  if (objectType === 'blob') {
    objectContent = createBlobContent(content);
  } else {
    // For other types, just use blob format for now
    const header = `${objectType} ${content.length}\0`;
    objectContent = Buffer.concat([Buffer.from(header), content]);
  }

  const sha = hashObject(objectContent);

  if (write) {
    const repoRoot = findRepoRoot();
    if (!repoRoot) {
      console.error('fatal: not a minigit repository');
      return 1;
    }
    writeObject(repoRoot, objectContent);
  }

  console.log(sha);
  return 0;
}
